go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -156,14 +287,21 @@ func (npw *nodePortWatcher) updateServiceFlowCache(service *kapi.Service, add bo
 
 // AddService handles configuring shared gateway bridge flows to steer External IP, Node Port, Ingress LB traffic into OVN
 func (npw *nodePortWatcher) AddService(service *kapi.Service) {
-
 	// don't process headless service or services that doesn't have NodePorts or ExternalIPs
 	if !util.ServiceTypeHasClusterIP(service) || !util.IsClusterIPSet(service) {
 		return
 	}
-	npw.updateServiceFlowCache(service, true)
-	npw.ofm.requestFlowSync()
-	addSharedGatewayIptRules(service)
+
+	name := ktypes.NamespacedName{Namespace: service.Namespace, Name: service.Name}
+	// Add services to npw struct so that the service object can be used in Endpoint Watcher functions
+	// The service can only be deleted from the npw struct upon Endpoint Deletion
+	// if the endpoint add event already created the service just exit
+	if _, exists := npw.services.Load(name); !exists {
+		npw.services.Store(name, service)
+		npw.updateServiceFlowCache(service, true, false)
+		npw.ofm.requestFlowSync()
+		addSharedGatewayIptRules(service, false, npw.nodeIPManager.addresses.List())
+	}
 }
 
 func (npw *nodePortWatcher) UpdateService(old, new *kapi.Service) {
```

Comment:
update is flawed. You could have this happen:
1. add endpoint event with host network backends
2. service update event and external ip changed

now you have flows for not host backends?

Really what I think you should do is make npw.services hold a struct with the service name and hasHostEPs bool. Then in each handler you will need to cover the possible cases.

Reply:
I need the actual service for `updateServiceFlowCache` though... So I think I just need to do a better job of handling the corner cases here.....

Reply:
Basically if any service Event happens that could change the ETP flows (spec ExternalTrafficPolicy, nodeport, External IP, etc) I need to just call AddEndpoint. 

It's safe because no endpoint event sould effect services

Reply:
So these are the flows we need to handle (in any order they could conflict)

2. Service Add AND Endpoint Add
- If Service Event arrives first just Make normal rules 
- Then If Endpoint event arrives ETP:local && hasHostNetworked backendts Make special rules 
------
- If Endpoint Event arrives before service event then.... Create in Map(which should stop the incoming service event from doing anything) and then create the service 

2. Service Update
- If relevant field in a service changes, add normal flows and then run an updateEndpoint flow to ensure we didn't miss any endpoint events during the service update 

3. Endpoint Update 
- If event came from Service Update (new == nil) then Get EP from the cache and make sure old == new if not return
- If externalTrafficPolicy:local and Endpoints changed and we changed state of having host endpoints then delete and readd
 
5. Service Delete AND Endpoint Delete 
Delete all flows for the service

Reply:
> update is flawed. You could have this happen:
add endpoint event with host network backends
service update event and external ip changed
now you have flows for not host backends?

Now the first add endpoint event should proceed normally, 

and the service update event will trigger another add endpoint event to ensure no events were missed